### ISSUE TITLE
Intermittent NullPointerException in 9.1.13.0

### ISSUE THREAD

**Issue by NeilSorensen** [jruby/jruby#5492]
### Environment

Version: ```jruby 9.1.13.0 (2.3.3) 2017-09-06 8e1c115 OpenJDK 64-Bit Server VM 25.191-b12 on 1.8.0_191-8u191-b12-0ubuntu0.16.04.1-b12 +indy +jit [linux-x86_64]```

We're running a custom logstash filter that uses the `json-schema` gem to validate that log files are in the correct format before sending them to our elasticsearch index. The code for our filter can be found at https://github.com/pluralsight/logstash-filter-json_schema.

### Expected Behavior

Calling ```JSON::Validator.validate``` should check the provided string against the provided schema, and return true if it matches.

### Actual Behavior

Approximately 10% of the time, when we restart logstash, we get NullPointerExceptions when we attempt to validate strings against a schema. This error occurs every time we call validate until we restart the process. From the backtrace (see below), it appears that there's some kind of race condition in JRuby that causes this problem. I hate to blame problems on a framework, but that's the only thing that makes sense to me at this point.

I recognize that this is more likely a problem with logstash than (strictly speaking) a problem with JRuby, but I'd appreciate confirmation that I'm following the backtrace correctly, as I'm neither a full time Ruby or Java developer.

Backtrace (I've only included the relevant portion, but can provide the full trace upon request):
```org.jruby.RubyString.getStringForPattern(RubyString.java:3741)
org.jruby.RubyString.asRegexpArg(RubyString.java:2405)
org.jruby.RubyString.subBangNoIter(RubyString.java:2445)
org.jruby.RubyString.sub_bang(RubyString.java:2398)
org.jruby.RubyString$INVOKER$i$sub_bang.call(RubyString$INVOKER$i$sub_bang.gen)
org.jruby.internal.runtime.methods.JavaMethod$JavaMethodOneOrTwoOrNBlock.call(JavaMethod.java:402)
uri_3a_classloader_3a_.META_minus_INF.jruby_dot_home.lib.ruby.stdlib.date.format.RUBY$method$_parse_day$0(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/date/format.rb:476)
uri_3a_classloader_3a_.META_minus_INF.jruby_dot_home.lib.ruby.stdlib.date.format.RUBY$method$_parse$0(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/date/format.rb:847)
uri_3a_classloader_3a_.META_minus_INF.jruby_dot_home.lib.ruby.stdlib.date.format.RUBY$method$_rfc3339$0(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/date/format.rb:1021)
uri_3a_classloader_3a_.META_minus_INF.jruby_dot_home.lib.ruby.stdlib.date.RUBY$method$rfc3339$0(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/date.rb:1840)```

**Comment by kares** [comment]
start of the stack-trace could be useful ... its not clear whether the NPE happened at the top level frame:
which would be `org.jruby.RubyString.asRegexpArg(RubyString.java:2405)`
also any reason to go at least to latest 9.1 patch release, or is that locked by your logstach version?

**Comment by craig-miskell-fluxfederation** [comment]
We are seeing the same, also triggered by logstash although on an S3 input.  The start of the stack trace is:
```
  Exception: Java::JavaLang::NullPointerException
  Stack: org.jruby.RubyString.getStringForPattern(RubyString.java:3741)
org.jruby.RubyString.asRegexpArg(RubyString.java:2405)
```
then follows as above.  To me one of the most interesting details is that lib/ruby/stdlib/date/format.rb:476 is:
    if str.sub!(/\b(#{Format::ABBR_DAYS.keys.join('|')})[^-\d\s]*/io, ' ')
and following the chain, the NPE is the regex arg to sub!, which is *clearly* not null, and never will be.

Also for context: our inputs were working fine until we added our 4th S3 bucket source.  The error will occur more readily than noted by Neil, but not every run; something like 80% of the time it'll happen as soon as the S3 calls start, but if not then things seem to survive.  Instinctively it looks like some sort of threading/race condition, triggered in our case by having multiple inputs running simultaneously.

I'll see if I can create a non-logstash trivial reproducer

**Comment by NeilSorensen** [comment]
Here's the full trace:
```org.jruby.RubyString.getStringForPattern(RubyString.java:3741)
org.jruby.RubyString.asRegexpArg(RubyString.java:2405)
org.jruby.RubyString.subBangNoIter(RubyString.java:2445)
org.jruby.RubyString.sub_bang(RubyString.java:2398)
org.jruby.RubyString$INVOKER$i$sub_bang.call(RubyString$INVOKER$i$sub_bang.gen)
org.jruby.internal.runtime.methods.JavaMethod$JavaMethodOneOrTwoOrNBlock.call(JavaMethod.java:402)
uri_3a_classloader_3a_.META_minus_INF.jruby_dot_home.lib.ruby.stdlib.date.format.RUBY$method$_parse_day$0(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/date/format.rb:476)
uri_3a_classloader_3a_.META_minus_INF.jruby_dot_home.lib.ruby.stdlib.date.format.RUBY$method$_parse$0(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/date/format.rb:847)
uri_3a_classloader_3a_.META_minus_INF.jruby_dot_home.lib.ruby.stdlib.date.format.RUBY$method$_rfc3339$0(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/date/format.rb:1021)
uri_3a_classloader_3a_.META_minus_INF.jruby_dot_home.lib.ruby.stdlib.date.RUBY$method$rfc3339$0(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/date.rb:1840)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.attributes.formats.date_time_v4.RUBY$method$validate$0(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/attributes/formats/date_time_v4.rb:8)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.attributes.format.RUBY$method$validate$0(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/attributes/format.rb:10)
org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:77)
org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:93)
org.jruby.ir.targets.InvokeSite.fail(InvokeSite.java:168)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.schema.validator.RUBY$block$validate$1(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/schema/validator.rb:24)
org.jruby.runtime.CompiledIRBlockBody.yieldDirect(CompiledIRBlockBody.java:156)
org.jruby.runtime.BlockBody.yield(BlockBody.java:114)
org.jruby.runtime.Block.yield(Block.java:165)
org.jruby.RubyHash$12.visit(RubyHash.java:1362)
org.jruby.RubyHash$12.visit(RubyHash.java:1359)
org.jruby.RubyHash.visitLimited(RubyHash.java:662)
org.jruby.RubyHash.visitAll(RubyHash.java:647)
org.jruby.RubyHash.iteratorVisitAll(RubyHash.java:1319)
org.jruby.RubyHash.each_pairCommon(RubyHash.java:1354)
org.jruby.RubyHash.each(RubyHash.java:1343)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.schema.validator.RUBY$method$validate$0(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/schema/validator.rb:22)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.schema.RUBY$method$validate$0(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/schema.rb:33)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.attributes.properties.RUBY$block$validate$1(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/attributes/properties.rb:32)
org.jruby.runtime.CompiledIRBlockBody.yieldDirect(CompiledIRBlockBody.java:156)
org.jruby.runtime.BlockBody.yield(BlockBody.java:114)
org.jruby.runtime.Block.yield(Block.java:165)
org.jruby.RubyHash$12.visit(RubyHash.java:1362)
org.jruby.RubyHash$12.visit(RubyHash.java:1359)
org.jruby.RubyHash.visitLimited(RubyHash.java:662)
org.jruby.RubyHash.visitAll(RubyHash.java:647)
org.jruby.RubyHash.iteratorVisitAll(RubyHash.java:1319)
org.jruby.RubyHash.each_pairCommon(RubyHash.java:1354)
org.jruby.RubyHash.each(RubyHash.java:1343)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.attributes.properties.RUBY$method$validate$0(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/attributes/properties.rb:14)
org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:77)
org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:93)
org.jruby.ir.targets.InvokeSite.fail(InvokeSite.java:168)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.schema.validator.RUBY$block$validate$1(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/schema/validator.rb:24)
org.jruby.runtime.CompiledIRBlockBody.yieldDirect(CompiledIRBlockBody.java:156)
org.jruby.runtime.BlockBody.yield(BlockBody.java:114)
org.jruby.runtime.Block.yield(Block.java:165)
org.jruby.RubyHash$12.visit(RubyHash.java:1362)
org.jruby.RubyHash$12.visit(RubyHash.java:1359)
org.jruby.RubyHash.visitLimited(RubyHash.java:662)
org.jruby.RubyHash.visitAll(RubyHash.java:647)
org.jruby.RubyHash.iteratorVisitAll(RubyHash.java:1319)
org.jruby.RubyHash.each_pairCommon(RubyHash.java:1354)
org.jruby.RubyHash.each(RubyHash.java:1343)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.schema.validator.RUBY$method$validate$0(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/schema/validator.rb:22)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.schema.RUBY$method$validate$0(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/schema.rb:33)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.validator.RUBY$method$validate$0(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/validator.rb:114)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.json_minus_schema_minus_2_dot_6_dot_2.lib.json_minus_schema.validator.RUBY$method$validate$0(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/json-schema-2.6.2/lib/json-schema/validator.rb:243)
usr.share.logstash.vendor.bundle.jruby.$2_dot_3_dot_0.gems.logstash_minus_filter_minus_json_schema_minus_0_dot_2_dot_1.lib.logstash.filters.json_schema.RUBY$method$filter$0(/usr/share/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-filter-json_schema-0.2.1/lib/logstash/filters/json_schema.rb:28)
usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.RUBY$method$do_filter$0(/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:143)
usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.RUBY$block$multi_filter$1(/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:162)
org.jruby.runtime.CompiledIRBlockBody.yieldDirect(CompiledIRBlockBody.java:156)
org.jruby.runtime.BlockBody.yield(BlockBody.java:114)
org.jruby.runtime.Block.yield(Block.java:165)
org.jruby.RubyArray.each(RubyArray.java:1734)
usr.share.logstash.logstash_minus_core.lib.logstash.filters.base.RUBY$method$multi_filter$0(/usr/share/logstash/logstash-core/lib/logstash/filters/base.rb:159)
usr.share.logstash.logstash_minus_core.lib.logstash.filter_delegator.RUBY$method$multi_filter$0(/usr/share/logstash/logstash-core/lib/logstash/filter_delegator.rb:44)
org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:103)
org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:163)
org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:200)
org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:161)
org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:314)
org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:73)
org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(Interpreter.java:132)
org.jruby.runtime.MixedModeIRBlockBody.commonYieldPath(MixedModeIRBlockBody.java:148)
org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:73)
org.jruby.runtime.Block.call(Block.java:124)
org.jruby.RubyProc.call(RubyProc.java:289)
org.jruby.internal.runtime.methods.ProcMethod.call(ProcMethod.java:63)
org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:204)
org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:200)
usr.share.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$method$filter_batch$0(/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:341)
usr.share.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$method$worker_loop$0(/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:320)
usr.share.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$method$worker_loop$0$__VARARGS__(/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb)
org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:77)
org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:93)
org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:145)
usr.share.logstash.logstash_minus_core.lib.logstash.pipeline.RUBY$block$start_workers$2(/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:286)
org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:145)
org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:71)
org.jruby.runtime.Block.call(Block.java:124)
org.jruby.RubyProc.call(RubyProc.java:289)
org.jruby.RubyProc.call(RubyProc.java:246)
org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:104)
java.lang.Thread.run(Thread.java:748)```

From what I can tell, logstash is locked to 9.1.13 right now. There's some effort (see https://github.com/elastic/logstash/issues/9345) to upgrade to 9.2, but I don't know how active that work is.

**Comment by headius** [comment]
@NeilSorensen Thanks for the longer trace! 

The code in question (in date/format.rb) appears to have been moved into Java for JRuby 9.2, so I can't really track whether the same NPE might still exist. I also looked at the RubyString methods where the NPE originates; in those places, the only way 9.2 could have a null would be if one were passed in from Ruby, which would indicate a larger bug somewhere (there should never be null references passed along Ruby call paths).

We are currently on the fence about doing a 9.1.18 release since 9.2 has stabilized well. There's a very good chance this issue has been resolved or aged out by the rework of date formatting into Java code.
